File: leetcode/lc-us/cpp/491.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    vector<int> path;
    vector<vector<int>> paths;
    vector<int> arr;

    void dfs(int cur, int prev) {
        if (cur == arr.size()) {
            if (path.size() >= 2) {
                paths.push_back(path);
            }
            return;
        }
        if (arr[cur] >= prev) {
            path.push_back(arr[cur]);
            dfs(cur + 1, arr[cur]);
            path.pop_back();
        }
        if (arr[cur] != prev) {
            dfs(cur + 1, prev);
        }
    }

public:
    vector<vector<int>> findSubsequences(vector<int> &nums) {
        arr = nums;
        dfs(0, INT32_MIN);
        return paths;
    }
};

File: leetcode/lc-us/py/491.py
from typing import List


class Solution:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        path, res = [], []

        def dfs(u: int) -> None:
            if u == len(nums):
                if len(path) >= 2:
                    res.append(path[:])
                return
            if not path or path[-1] < nums[u]:
                dfs(u + 1)

                path.append(nums[u])
                dfs(u + 1)
                path.pop()
            elif path[-1] == nums[u]:
                path.append(nums[u])
                dfs(u + 1)
                path.pop()
            else:
                dfs(u + 1)

        dfs(0)
        return res
